[SVLS-6527] Step Functions Distributed Map (#28469)
DataDog/documentation: content/en/serverless/step_functions/distributed-maps.md

---
title: Tracing Distributed Map states
---

> This page explains our support for tracing your Map states in AWS Step Functions set to [Distributed mode][1].

Reviewer: ```suggestion
In AWS Step Functions, you can set up a large-scale parallel workload by including a [`Map` state in _Distributed_ mode][1]. Datadog supports tracing your Distributed Map states.
```

{{< img src="serverless/step_functions/distributed_map.png" alt="A visualization of a Step Function with a Distributed Map state." style="width:100%;" >}}

> ## Trace Merging

Reviewer: ```suggestion
## Trace merging
```

> To ensure that the child workflows correctly parent the Distributed map, you must enable the "Export Map state results to Amazon S3" option. Learn more about [exporting to S3][2].

Reviewer: I'm not quite sure what is meant by "child workflows correctly parent the Distributed map"... can we change this to:

```suggestion
To ensure that child workflows are correctly linked to their parents, enable the **Export Map state results to Amazon S3** option for your Step Function. For more information, see [Exporting to Amazon S3][2].
```

Is this what you meant?

## Limitations

> - We don't support parenting for failed child workflows. Meaning they'll be in the same trace as the parent Distributed map but as an orphan.

Reviewer: ```suggestion
- Datadog does not support parenting for failed child workflows. You can find these failed child workflows as orphans in the same trace as the parent Distributed Map.
```

> - We cannot propagate trace context if there is another Step Function or a Lambda upstream from the Distributed map's Step Function. The link will be broken and the child workflow will be in its own trace.

Reviewer: ```suggestion
- Datadog cannot propagate trace context if there is another Step Function or a Lambda upstream from the Distributed Map's Step Function. The link is broken, and the child workflow is in its own trace.
```

> - Currently, the executions from the child map run will be in the same invocation table as the parent Step Function. The child table in the Step Functions page will be empty as a result.

Reviewer: ```suggestion
- Executions from a child map run are in the same invocation table as the parent Step Function. As a result, the child table in the Step Functions page is empty.
```

## Troubleshooting

> - If trace merging is broken because there is another upstream service, you can use the `fan_out_children_trace_id` span attribute in the Distributed map span to look for the child workflow traces.

Reviewer: ```suggestion
- If trace merging is broken because there is another upstream service, you can use the `fan_out_children_trace_id` span attribute in the Distributed Map span to look for the child workflow traces.
```

---
title: Tracing Distributed Map States
---

In AWS Step Functions, you can set up a large-scale parallel workload by including a [`Map` state in _Distributed_ mode][1]. Datadog supports tracing your Distributed Map states.

{{< img src="serverless/step_functions/distributed_map.png" alt="A visualization of a Step Function with a Distributed Map state." style="width:100%;" >}}

## Set up trace merging

1. To ensure that child workflows are correctly linked to their parents, enable the **Item Batching** option for your Distributed Map state. For more information, see [ItemBatcher][2].

2. Your State Machine definition must use [JSONata][4] as the query language. To enable this, set your definition's top-level `QueryLanguage` field to `JSONata`.

3. On the Distributed Map state, set `_datadog` in the `BatchInput` field as follows:

   {{< highlight json "hl_lines=4-4" >}}
   "ItemBatcher": {

>      "MaxItemsPerBatch": 100,

Reviewer 2: I think a note specifically calling this number out as user configurable would increase clarity

Author: Set the value to N to communicate that
